File: app/src/main/java/com/example/waveplayer/random_playlist/AudioUri.java
package com.example.waveplayer.random_playlist;

import android.content.ContentUris;
import android.content.Context;
import android.media.MediaMetadataRetriever;
import android.net.Uri;
import android.provider.MediaStore;
import android.util.Log;

import androidx.annotation.Nullable;

import java.io.Serializable;
import java.util.Random;

public final class AudioUri implements Comparable<AudioUri>, Serializable {

    transient static final String TAG = "AudioURI";

    transient private Uri uri;

    private final NestedProbMap nestProbMap = new NestedProbMap();

    public final String displayName;

    public final String artist;

    public final String title;

    public final long id;

    public Uri getUri() {
        Log.v(TAG, "getUri started");
        if (uri == null) {
            this.uri = ContentUris.withAppendedId(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI, id);
        }
        Log.v(TAG, "getUri ended");
        return uri;
    }

    public static Uri getUri(Long songID) {
        return ContentUris.withAppendedId(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI, songID);
    }

    private int duration = -1;

    public int getDuration(Context context) {
        Log.v(TAG, "getDuration started");
        if (duration == -1) {
            MediaMetadataRetriever mediaMetadataRetriever = new MediaMetadataRetriever();
            mediaMetadataRetriever.setDataSource(context, getUri());
            String time = mediaMetadataRetriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION);
            if (time == null) {
                time = "00:00:00";
            }
            this.duration = Integer.parseInt(time);
            mediaMetadataRetriever.release();
        }
        Log.v(TAG, "getDuration ended");
        return duration;
    }

    public AudioUri(String displayName, String artist, String title, long id) {
        //Log.v(TAG, "AudioURI constructing");
        this.displayName = displayName;
        this.artist = artist;
        this.title = title;
        this.id = id;

        // Log.v(TAG, "AudioURI constructed");
    }

    public boolean shouldPlay(Random random) {
        return nestProbMap.outcome(random);
    }

    public boolean bad(double percent) {
        return nestProbMap.bad(percent);
    }

    public boolean good(double percent) {
        return nestProbMap.good(percent);
    }

    public void clearProbabilities() {
        nestProbMap.clearProbabilities();
    }

        @Override
    public int compareTo(AudioUri o) {
            return title.compareTo(o.title);
    }

    @Override
    public boolean equals(@Nullable Object obj) {
        //Log.v(TAG, "equals start");
        if (uri == null) {
            this.uri = ContentUris.withAppendedId(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI, id);
        }
        // Log.v(TAG, "equals end");
        return obj instanceof AudioUri && uri.equals(((AudioUri) obj).getUri());
    }

    @Override
    public int hashCode() {
        //Log.v(TAG, "hashCode start");
        if (uri == null) {
            this.uri = ContentUris.withAppendedId(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI, id);
        }
        //Log.v(TAG, "hashCode end");
        return uri.toString().hashCode();
    }
}

File: app/src/main/java/com/example/waveplayer/random_playlist/AudioUri.kt
package com.example.waveplayer.random_playlist

import android.content.ContentUris
import android.content.Context
import android.media.MediaMetadataRetriever
import android.net.Uri
import android.provider.MediaStore
import java.io.*
import java.util.*

class AudioUri(
        val displayName: String,
        val artist: String,
        val title: String,
        val id: Long
        ) : Comparable<AudioUri>, Serializable {

    val uri: Uri by lazy{
        getUri(id)
    }

    private val nestProbMap: NestedProbMap = NestedProbMap()

    private var duration: Int = -1
    fun getDuration(context: Context): Int {
        if (duration == -1) {
            val mediaMetadataRetriever = MediaMetadataRetriever()
            mediaMetadataRetriever.setDataSource(context, uri)
            var time = mediaMetadataRetriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION)
            if (time == null) {
                time = "00:00:00"
            }
            duration = time.toInt()
            mediaMetadataRetriever.release()
        }
        return duration
    }

    fun shouldPlay(random: Random): Boolean {
        return nestProbMap.outcome(random)
    }

    fun bad(percent: Double): Boolean {
        return nestProbMap.bad(percent)
    }

    fun good(percent: Double): Boolean {
        return nestProbMap.good(percent)
    }

    fun clearProbabilities() {
        nestProbMap.clearProbabilities()
    }

    override fun compareTo(other: AudioUri): Int {
        return title.compareTo(other.title)
    }

    override fun equals(other: Any?): Boolean {
        return other is AudioUri && uri == other.uri
    }

    override fun hashCode(): Int {
        return uri.toString().hashCode()
    }

    companion object {
        fun getAudioUri(context: Context, songID: Long): AudioUri? {
            val file = File(context.filesDir, songID.toString())
            if (file.exists()) {
                try {
                    context.openFileInput(songID.toString()).use { fileInputStream ->
                        ObjectInputStream(fileInputStream).use { objectInputStream ->
                            return objectInputStream.readObject() as AudioUri } }
                } catch (e: FileNotFoundException) {
                    e.printStackTrace()
                } catch (e: IOException) {
                    e.printStackTrace()
                } catch (e: ClassNotFoundException) {
                    e.printStackTrace()
                }
            }
            return null
        }

        fun getUri(songID: Long): Uri {
            return ContentUris.withAppendedId(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI, songID)
        }

    }
}